### PR TITLE
Article added for N-Queens Problem

### PR DESCRIPTION
**Pull request by DEEPIKA-SUK**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

YES

### Description

Added the article for N-Queens Problem. The basic layout for the article is:
What is the N Queens Problem?
Explanation for N Queen
Algorithmic Steps
Code in C++ and Python
Sample I/O
Time & Space Complexity

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [ ] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

Issue: Content: N-Queens #20

### CODE REVIEW

**Comment by DEEPIKA-SUK** [comment]
@HarshCasper Kindly review my PR.

File: docs/backtracking/n-queens.mdx [hunk 1]
```diff
@@ -0,0 +1,221 @@
+---
+title: N Queens
+slug: /backtracking/n-queens
+---
+
+## Introduction
+
+The N Queen problem is placing N chess queens on an N x N chessboard such that no two queens can attack each other.
+The chess queens can attack in any of the following directions: horizontal, vertical, and diagonal.
+Output is the N x N binary matrices showing possible positions for queens such that they cannot attack each other, where 1 represents the correct place for the queen. If no such valid solution exists, then output is `false`.
```
Comment (HarshCasper) [inline docs/backtracking/n-queens.mdx:10]: Looks too similar to [this](https://www.geeksforgeeks.org/n-queen-problem-backtracking-3/). Kindly don't plagiarize.

File: docs/backtracking/n-queens.mdx [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+## Explaination
+
+The backtracking approach is used to solve this problem. Queens are placed one by one in different columns, starting from the leftmost column. While placing a queen in a column, we see if that position is safe or not. If it's safe, mark this row and column as part of the solution else, backtrack and return false.
```
Comment (HarshCasper) [inline docs/backtracking/n-queens.mdx:14]: Copied from [here](https://www.geeksforgeeks.org/n-queen-problem-backtracking-3/)

File: docs/backtracking/n-queens.mdx [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+## Algorithm
+
+- Step 1: Start from the leftmost position.
+- Step 2: If all queens are placed, then return true.
+- Step 3: Try all rows in the column. For each row, do the following:
+  - a) If this position is safe for the queen to be placed, then mark this [row, column] as part of the solution and see if this position leads to a solution or not using recursion.
+  - b) If that position leads to a solution, then return true else unmark [row, column] this position
+    i.e., backtrack and go to a) and repeat the same procedure for other rows.
+- Step 4: If we don't get any solution after trying all rows, then return false.
```
Comment (HarshCasper) [inline docs/backtracking/n-queens.mdx:24]: Copied from [here](https://www.geeksforgeeks.org/printing-solutions-n-queen-problem/)

**Review by HarshCasper — CHANGES_REQUESTED** [review]
Kindly remove the plagiarism or else I will not be able to accept the PR.

**Comment by DEEPIKA-SUK** [comment]
@HarshCasper I have not copied, I have read several articles and then written everything in my own words. Still I will do the necessary changes.

**Comment by HarshCasper** [comment]
> @HarshCasper I have not copied, I have read several articles and then written everything in my own words. Still I will do the necessary changes.

Yeah, please. I understand your point but here some of the parts are too obvious. We don't want to bring any negative impression on the project.

**Comment by DEEPIKA-SUK** [comment]
@HarshCasper I have made the necessary changes. Kindly review my PR and let me know if any further changes are to be made.

File: docs/backtracking/n-queens.mdx [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+  |       |       |       |       |
+  |-------+-------+-------+-------+
+  |       |       |       |       |
+2 |   X   |       |       |       |
+  |       |       |       |       |
+  |-------|-------|-------|-------|
+  |       |       |       |       |
+3 |       |       |   X   |       |
+  |       |       |       |       |
+  |-------|-------|-------|-------|
```
Comment (HarshCasper) [inline docs/backtracking/n-queens.mdx:247]: This is not even a part of the output log. Why is it here?

File: docs/backtracking/n-queens.mdx [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+
+```
+
+## Complexity Analysis
+
+- `N` is the size of the Chessboard.
+- Space Complexity : O(N^2)
+- Time Complexity : upperbounded by O(N^N)
```
Comment (HarshCasper) [inline docs/backtracking/n-queens.mdx:255]: Keep the values in backticks

File: docs/backtracking/n-queens.mdx [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+    }
+}
+
+int main() {
+	int n;
+	cin >> n;
+
+	bool board[N][N];
+    bool cols[N];
+    bool ndiag[2*N-1], rdiag[2*N-1];
```
Comment (HarshCasper) [inline docs/backtracking/n-queens.mdx:117]: Indent the Code properly

File: docs/backtracking/n-queens.mdx [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+    // initialize that no column has a queen
+	for(int i = 0; i < n; i++)
+		cols[i] = false;
+
+	for(int i = 0; i < 2*n-1; i++) {
+		ndiag[i] = false;
+		rdiag[i] = false;
+	}
+
+	solve(board, cols, ndiag, rdiag, 0, "", n);
```
Comment (HarshCasper) [inline docs/backtracking/n-queens.mdx:133]: The input and output statements of C++ and Python don't match. Kindly fix them up

File: docs/backtracking/n-queens.mdx [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+*---*---*---*---*
+|   |   |   | X |
+*---*---*---*---*
+|   | X |   |   |
+*---*---*---*---*
+|   |   |   |   |
+*---*---*---*---*
+```
+
+Now the 4th queen cannot be placed in the 4th row as there exists no safe cell such that no two queens can attack each other. So we will return false and backtrack to the previous row and change the queen's position. We will repeat the same procedure untill we find the positions for the queens.
```
Comment (HarshCasper) [inline docs/backtracking/n-queens.mdx:42]: Keep a line break after this

File: docs/backtracking/n-queens.mdx [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+Therefore, the answer for N=4 is:
```
Comment (HarshCasper) [inline docs/backtracking/n-queens.mdx:43]: Keep `N=4` in backticks

File: docs/backtracking/n-queens.mdx [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+
+## Introduction
+
+In this problem, we are provided an N x N chessboard, and we have to find if there exists any way to place N chess queens such that no two queens can attack each other. The possible directions where the queen can move and attack are horizontal, vertical, and diagonal. If there exists any way then the output is going to be the corresponding positions of N queens else `false` will be returned.
+
+## Explaination
+
+This problem can be solved using `backtracking`. Each row and column will contain only one queen. We will start from the first cell of the chessboard. We will move row-wise, and for each row, we will see if there exists a column where the queen can be placed safely, i.e., no two queens are at an attacking position, we will include that cell in solution and will proceed further, and if we didn't find any such cell, then we will return false and backtrack to previous cell, change the queen's position.
+
+Let's take an example, N=4, `X` represents chess queen.
```
Comment (HarshCasper) [inline docs/backtracking/n-queens.mdx:14]: Keep `N=4` in backticks

File: docs/backtracking/n-queens.mdx [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+slug: /backtracking/n-queens
+---
+
+## Introduction
+
+In this problem, we are provided an N x N chessboard, and we have to find if there exists any way to place N chess queens such that no two queens can attack each other. The possible directions where the queen can move and attack are horizontal, vertical, and diagonal. If there exists any way then the output is going to be the corresponding positions of N queens else `false` will be returned.
+
+## Explaination
+
+This problem can be solved using `backtracking`. Each row and column will contain only one queen. We will start from the first cell of the chessboard. We will move row-wise, and for each row, we will see if there exists a column where the queen can be placed safely, i.e., no two queens are at an attacking position, we will include that cell in solution and will proceed further, and if we didn't find any such cell, then we will return false and backtrack to previous cell, change the queen's position.
```
Comment (HarshCasper) [inline docs/backtracking/n-queens.mdx:12]: Keep backtracking in bolds

Comment (HarshCasper) [inline docs/backtracking/n-queens.mdx:12]: Don't use short forms like i.e. Replace with `that is`

Comment (HarshCasper) [inline docs/backtracking/n-queens.mdx:12]: `We will move row-wise, and for each row, we will see if there exists a column where the queen can be placed safely, i.e., no two queens are at an attacking position, we will include that cell in solution and will proceed further, and if we didn't find any such cell, then we will return false and backtrack to previous cell, change the queen's position.`

Break this into smaller lines.

**Review by HarshCasper — CHANGES_REQUESTED** [review]

**Comment by DEEPIKA-SUK** [comment]
@HarshCasper Kindly review now, I have made all the changes.

**Review by HarshCasper — APPROVED** [review]

**Pull request merged**